### ISSUE TITLE
Kitchen Sink Note Example Freezes on Key Press Hold

### ISSUE THREAD

**Issue by leovigna** [TanStack/router#1250]
## TLDR
* TextArea input freezes when user holds down key due to excessive `navigate` calls
* Solution is to debounce `notes` variable by 50ms 

## Bug
The kitchen sink note input is a good example of using TanStack Router with input forms.
However, currently the example implementation suffers from a bug when a user holds down a key. This is because it triggers `navigate` instantly which triggers many renders and interrupts the input. 

The example should implement the proper debounce behaviour such that other developers are not discouraged from storing form input in the url using TanStack.

Note: This bug may also be applicable to other examples.

## Solution
Simply debounce the `notes` variable in  [dashboard/invoices/$invoiceId](https://github.com/TanStack/router/blob/main/examples/react/kitchen-sink-file-based/src/routes/dashboard.invoices.%24invoiceId.tsx) and hold down a key.

Example solution using [usehooks.com/usedebounce](https://usehooks.com/usedebounce)
```tsx
// /examples/react/kitchen-sink-file-based/src/routes/dashboard.invoices.$invoicedId.tsx

//...
import { useDebounce } from '@uidotdev/usehooks'
//...
function InvoiceComponent() {
  //...
  const [notes, setNotes] = React.useState(search.notes ?? '')

  // Debounce `notes` by 50ms
  // Holding down a key while inputing a note will NOT trigger navigate
  // Also see https://usehooks.com/usedebounce
  const debouncedNotes = useDebounce(notes, 50);

  React.useEffect(() => {
    navigate({
      search: (old) => ({
        ...old,
        notes: debouncedNotes ? debouncedNotes : undefined,
      }),
      replace: true,
    })
  }, [debouncedNotes])
  
  return (
  //...
  )
}
```

I can make a PR for this and also see if any other examples suffer from same problem.

### Your Example Website or App

https://tanstack.com/router/latest/docs/framework/react/examples/kitchen-sink-file-based

### Steps to Reproduce the Bug or Issue

* Clone TanstackRouter
* Try the `kitchen-sink-file-based` example
* Go to http://localhost:3001/dashboard/invoices/1?showNotes=true
* Enter a note and hold down a key, watch as input gets interrupted
* See code here [dashboard/invoices/$invoiceId](https://github.com/TanStack/router/blob/main/examples/react/kitchen-sink-file-based/src/routes/dashboard.invoices.%24invoiceId.tsx) and hold down a key.

### Expected behavior

As a user, I expect a smooth input in the text area and the URL to change when possible to save my draft.

### Platform

All platforms

**Comment by schiller-manuel** [comment]
not sure if this is really needed... the kitchen sink example does a lot of things a real application would not do, so why start here?

**Comment by leovigna** [comment]
The docs don't provide a guide about using TanStack Router to store user input data so the examples are what I looked at for guidance on whether this was a possible pattern (I'd seen it on some earlier videos exploring this example as well).

This bug isn't very intuitive (at least to me) and if not for digging deeper and finding this solution, I may have thought that form state cannot be stored in the URL.

I agree that fixing the bug in the example is probably not enough (as many people might miss it) and a section in the docs about using TanStack Router for storing input data might be more effective.

**Comment by SeanCassiere** [comment]
> I agree that fixing the bug in the example is probably not enough (as many people might miss it) and a section in the docs about using TanStack Router for storing input data might be more effective.

The problem comes where in the documentation would this be covered?

I'm thinking under the *navigation* page. My only concern is that this makes an already lengthy page, even longer.

**Comment by leovigna** [comment]
Looking at the docs, my feedback as while they are very in-depth and detailed, there is no clear separation between features & use cases. What is the difference? For me features describe how Tanstack Router works (eg. Route Matching, Outlets, PathParams) and use cases refers to what common patterns and how to build with Tanstack Router (eg. Authenticated Routes, Loading Data...)

For now I think the [Search Params OG State Manager](https://tanstack.com/router/latest/docs/framework/react/guide/search-params#search-params-the-og-state-manager) page seems most relevant but maybe a dedicated page on the "State Manager" concept could also be useful.

I'd encourage checking out this discussion I've created on thinking about Best Practices
https://github.com/TanStack/router/discussions/1249

**Comment by leovigna** [comment]
Providing more feedback on docs & examples, I've also had some challenges issues with the [authenticated-routes](https://tanstack.com/router/latest/docs/framework/react/guide/authenticated-routes) guide.

I tried adapting it to my auth provider Clerk and its [useUser()](https://clerk.com/docs/references/react/use-user) hook and therefore looked into the [router-context](https://tanstack.com/router/latest/docs/framework/react/guide/router-context) docs for dependency injection. The issue I had though was none of those 2 pages mentioned `router.invalidate()` which led to issues with the context not being up to date. 

Finally the auth in kitchen sink did provide more help (that's where I saw `router.invalidate()` though the example is not a reactive one (auth is just an object) and calls `react.invalidate()` just on button click instead of watching for actual auth state changes.
My end goal was to have something like the code below at the top of my app, and then manage everything else through the router context.
```tsx
function App() {
  //...
  React.useEffect(() => {
      router.invalidate()
    }, [auth.status])
  //...
}
```

**Comment by SeanCassiere** [comment]
Following the precedent set in #1392, I'm closing this issue as the examples are demos and do not need to be overly complicated.

Open to any PRs about improving the documentation, or if you want to incorporate some aspects from your [discussion](https://github.com/TanStack/router/discussions/1249) added in.
